### PR TITLE
Object Detection From Webcam Stream Guide

### PR DESCRIPTION
**Pull request by freddyaboulton**
## Description

Adds an a guide on detecting objects from the webcam stream. Also adds a "Streaming" section to the guides.

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

File: guides/07_streaming/02_object-detection-from-webcam.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
+# Object Detection from a Webcam Stream
```
Comment (freddyaboulton) [inline guides/07_streaming/02_object-detection-from-webcam.md:1]: This is basically the only change in this PR apart from the demo. 

I think this is a good starting point. As discussed earlier today, I think we can improve some things internally to improve the experience on spaces.

File: demo/yolov10_webcam_stream/run.py
Comment (abidlabs) [inline demo/yolov10_webcam_stream/run.py:1]: This will need a requirements.txt file yeah?

File: demo/yolov10_webcam_stream/run.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+def yolov10_inference(image, conf_threshold):
+    width, _ = image.size
+    import time
+
+    start = time.time()
+    results = model.predict(source=image, imgsz=width, conf=conf_threshold)
+    end = time.time()
+    annotated_image = results[0].plot()
+    print("time", end - start)
```
Comment (abidlabs) [inline demo/yolov10_webcam_stream/run.py:16]: Do we want to keep this time logic?

File: demo/yolov10_webcam_stream/run.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    </h1>
+    """
+    )
+    gr.HTML(
+        """
+        <h3 style='text-align: center'>
+        <a href='https://arxiv.org/abs/2405.14458' target='_blank'>arXiv</a> | <a href='https://github.com/THU-MIG/yolov10' target='_blank'>github</a>
+        </h3>
+        """
+    )
```
Comment (abidlabs) [inline demo/yolov10_webcam_stream/run.py:38]: I feel like a one-liner `gr.Markdown()` is sufficient? We don't really need the arxiv link etc.

File: guides/07_streaming/02_object-detection-from-webcam.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+```
+
+We will use the `plot` method to draw a bounding box around each detected object. YoloV10 asses images are in the BGR color format, so we will flip them to be in the expected RGB format of web browsers.
+
+## The Gradio Demo
+
+The Gradio demo will be pretty straightforward but we'll do a couple of things that are specific to streaming:
+
+* The user's webcam will be both an input and an output. That way, the user will only see their stream with the detected objects.
+* We'll use the `time_limit` and `stream_every` parameters of the `stream` event. The `time_limit` parameter will mean that we'll process each user's stream for that amount of time. The `stream_every` function will control how frequently the webcam stream is sent to the server.
```
Comment (abidlabs) [inline guides/07_streaming/02_object-detection-from-webcam.md:45]:
```suggestion
* We'll use the `time_limit` and `stream_every` parameters of the `stream` event. The `time_limit` parameter will mean that we'll process each user's stream for that amount of time. In a multi-user setting, such as on Spaces, this means that after this period of time, the next user in the queue will be able to use the demo. The `stream_every` function will control how frequently the webcam stream is sent to the server.
```

**Comment by abidlabs** [comment]
This looks great @freddyaboulton. I agree about creating a separate section on Streaming for visbility. However, right now, it only contains example applications. What do you think about moving the two streaming guides (streaming inputs and streaming outputs) to this section as well?

**Comment by abidlabs** [comment]
For the demo, I experienced a ton of lag, but it could be because I was on a pretty bad wifi connection. Nevertheless, could you try the demo on a slower connection through the browser and see how it works for you? If there's anything we can do to optimize the experience for slow connections, I think that'd be great

image

If the deme is running well for you, LGTM, everything else is pretty minor

File: demo/yolov10_webcam_stream/run.py (hunk 2 above)
Comment (freddyaboulton) [reply demo/yolov10_webcam_stream/run.py:16]: I think it helps to debug

File: demo/yolov10_webcam_stream/run.py (hunk 3 above)
Comment (freddyaboulton) [reply demo/yolov10_webcam_stream/run.py:38]: Yes

**Comment by freddyaboulton** [comment]
Thank you @abidlabs for the review! Yes the lag on spaces is on my radar.

**Comment by abidlabs** [comment]
Sweet, I'll take another pass very shortly

**Comment by freddyaboulton** [comment]
>  What do you think about moving the two streaming guides (streaming inputs and streaming outputs) to this section as well?

I like where the streaming guides currently are. What I decided to do is link from the basic streaming guides to the "end-to-end" example guides.

File: guides/04_additional-features/02_streaming-outputs.md [hunk 5]
```diff
@@ -67,4 +67,8 @@ gr.Interface(keep_repeating,
              gr.Video(sources=["webcam"], format="mp4"),
              gr.Video(streaming=True, autoplay=True)
 ).launch()
-```
\ No newline at end of file
+```
+
+## End-to-End Examples
+
+For an end-to-end example of streaming media, see the object detection from video [guide](/main/guides/object-detection-from-video) or the streaming ai generated audio with [transformers](https://huggingface.co/docs/transformers/index) [guide](/main/guides/streaming-ai-generated-audio).
```
Comment (abidlabs) [inline guides/04_additional-features/02_streaming-outputs.md:74]:
```suggestion
For an end-to-end example of streaming media, see the object detection from video [guide](/main/guides/object-detection-from-video) or the streaming AI-generated audio with [transformers](https://huggingface.co/docs/transformers/index) [guide](/main/guides/streaming-ai-generated-audio).
```

File: guides/07_streaming/02_object-detection-from-webcam.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+```
+
+
+## Conclusion
+
+You can check out our demo hosted on Hugging Face Spaces [here](https://huggingface.co/spaces/gradio/YOLOv10-webcam-stream). 
+
+It is also embedded on this page below
+
+$demo_YOLOv10-webcam-stream
```
Comment (abidlabs) [inline guides/07_streaming/02_object-detection-from-webcam.md:93]:
```suggestion
$demo_yolov10_webcam_stream
```

**Review by abidlabs — APPROVED** [review]
LGTM @freddyaboulton! Great demos and docs

File: guides/07_streaming/02_object-detection-from-webcam.md (hunk 6 above)
Comment (freddyaboulton) [reply guides/07_streaming/02_object-detection-from-webcam.md]: Good catch!

**Comment by freddyaboulton** [comment]
Thank you for catching the typos!

**Pull request merged**